Jenkinsfile

```diff
@@ -1,142 +1,170 @@
-#!groovy
-
-def tryStep(String message, Closure block, Closure tearDown = null) {
-  try {
-    block()
+pipeline {
+  agent any
+  options {
+    timeout(time: 5, unit: 'DAYS')
   }
-  catch (Throwable t) {
-    slackSend message: "${env.JOB_NAME}: ${message} failure ${env.BUILD_URL}", channel: '#ci-channel', color: 'danger'
 
-    throw t
-  }
-  finally {
-    if (tearDown) {
-      tearDown()
-    }
+  environment {
+    COMMIT_HASH = GIT_COMMIT.substring(0, 8)
+    PROJECT_PREFIX = "${BRANCH_NAME}_${COMMIT_HASH}_${BUILD_NUMBER}_"
+    IMAGE_BASE = "repo.secure.amsterdam.nl/mijnams/mijnamsterdam"
+    IMAGE_BUILD = "${IMAGE_BASE}:${BUILD_NUMBER}"
+    IMAGE_ACCEPTANCE = "${IMAGE_BASE}:acceptance"
+    IMAGE_PRODUCTION = "${IMAGE_BASE}:production"
+    IMAGE_TEST = "${IMAGE_BASE}:test"
+    IMAGE_LATEST = "${IMAGE_BASE}:latest"
   }
-}
 
+  stages {
 
-String BRANCH = "${env.BRANCH_NAME}"
-String DEV_TARGET = "acceptance"
-String PROD_TARGET = "production"
-
-node {
-  stage("Checkout") {
-    checkout scm
-  }
-
-  if (BRANCH != "test-acc") {
-    stage("Test") {
-      tryStep "test", {
-        sh "docker-compose -p mijn_amsterdam_frontend -f docker-compose.yml build && " +
-        "docker-compose -p mijn_amsterdam_frontend -f docker-compose.yml run --rm test"
+    stage('Unit tests') {
+      when { not { branch 'test' } } // Skip unit tests when pushing directly to test (for speed)
+      options {
+        timeout(time: 30, unit: 'MINUTES')
+      }
+      environment {
+        PROJECT = "${PROJECT_PREFIX}unit"
+      }
+      steps {
+        sh "docker-compose -p ${PROJECT} up --build --exit-code-from test-unit test-unit"
+      }
+      post {
+        always {
+          sh "docker-compose -p ${PROJECT} down -v || true"
+        }
       }
     }
-  }
-}
-
-if (BRANCH == "test") {
-  DEV_TARGET = "test"
-}
 
-if (BRANCH == "master") {
-  stage('Waiting for approval') {
-    input "Deploy master to ACC?"
-  }
-}
+    // TEST
 
-if (BRANCH == "master" || BRANCH == "test-acc" || BRANCH == "test") {
-
-  stage("Build image (test/acc)") {
-    tryStep "build", {
-      docker.withRegistry("${DOCKER_REGISTRY}", 'docker-registry') {
-      def image = docker.build("mijnams/mijnamsterdam:${env.BUILD_NUMBER}", "--build-arg PROD_ENV=${DEV_TARGET} --build-arg http_proxy=${JENKINS_HTTP_PROXY_STRING} --build-arg https_proxy=${JENKINS_HTTP_PROXY_STRING} .")
-      image.push()
+    stage('Build TEST') {
+      when { branch 'test' }
+      options {
+        timeout(time: 30, unit: 'MINUTES')
+      }
+      steps {
+        sh "docker build -t ${IMAGE_BUILD} " +
+          "--shm-size 1G " +
```

Comment:
Do we need this?

Reply:
Increasing memory to speed up task execution within the containers